xnmt/decoder.py

```diff
@@ -113,10 +118,16 @@ def calc_loss(self, context, ref_action):
     scores = self.get_scores(context)
     # single mode
     if not xnmt.batcher.is_batched(ref_action):
-      return dy.pickneglogsoftmax(scores, ref_action)
-    # minibatch mode
+      log_loss = dy.pickneglogsoftmax(scores, ref_action)
+
+    else: # minibatch mode
+      log_loss = dy.pickneglogsoftmax_batch(scores, ref_action)
+
+    if self.label_smoothing == 0.0:
+      return log_loss
     else:
-      return dy.pickneglogsoftmax_batch(scores, ref_action)
+      ce_term = (1 - self.label_smoothing) * (dy.exp(-log_loss))
+      return -dy.log(ce_term + self.label_smoothing_term)
```

Comment:
Hi. I've implemented label smoothing in my own code and I think the implementation should be like:

```python
ce_term = dy.pick_batch(log_loss, ref_action)
label_smoothing_term = dy.mean_elems(log_loss)
final_loss = (1 - self.label_smoothing) * ce_term + self.label_smoothing * label_smoothing_term
```

Basically you are just changing the target distribution so you don't need to leave log-space to compute the new cross entropy. Also in this case mean_elems is probably faster than dividing by the vocabulary size.

I think what you're doing now is replacing  H(q, p) by H(q, (1- eps)p + eps u) instead of H((1-eps)q + eps u, p) (using the notations of https://arxiv.org/pdf/1512.00567.pdf

Reply:
Hi Paul,
Thanks a lot for the helpful comments. I have now re-implemented the label smoothing functionality as you have suggested.

I observe that the speed: words/sec have dropped by almost 30-40% when using label smoothing (mayne due to computing the softmax over all the words in vocab). Did you also observe something similar.

Reply:
I got something around a 5% slow-down... Maybe this is because you seem to be computing the softmax twice (in pickneglogsoftmax for the normal loss term and in log_softmax for the label smoothing term). 

Whenever I used label smoothing I would compute `log_prob=log_softmax(score)` and then do all operation on this expression (ie not doing `pickneglogsoftmax`, but `pick` on `log_prob`).